File: node/fidelity-helper/src/fidelity-models.ts
export interface GetTransactionsReqModel {
  operationName: string;
  variables: {
    isNewOrderApi: boolean;
    isSupportCrypto: boolean;
    hideDCOrders: boolean;
    acctIdList: string;
    acctDetailList: {
      acctNum: string;
      acctType: string;
      name: string;
    }[];
    searchCriteriaDetail: {
      timePeriod: number;
      txnCat: null;
      viewType: string;
      histSortDir: string;
      acctHistSort: string;
      hasBasketName: boolean;
      txnFromDate: string;
      txnToDate: string;
    };
  };
  query: string;
}

export interface GetTransactionsRespHistoryModel {
  acctNum: string;
  amount: string;
  date: string;
  description: string;
  intradayInd: boolean;
  orderNumber: string;
}

export interface GetTransactionsRespModel {
  data: {
    getTransactions: {
      historys: GetTransactionsRespHistoryModel[];
    };
  };
}

export interface GetAccountsRespModel {
  data: {
    getContext: {
      sysStatus: {
        backend: {
          account: string;
        };
      };
      person: {
        assets: {
          acctNum: string;
          acctType: string;
          preferenceDetail: {
            name: string;
          };
        }[];
      };
    };
  };
}

File: python/fidelity-helper/src/fidelity_helper/fidelity_models.py
"""Fidelity request and response objects. Use with caution."""

# ruff: noqa: N815  # mixed-case-variable-in-class-scope
# ruff: noqa: D101  # Missing docstring in public class

from pydantic import BaseModel, ConfigDict


class AcctDetailModel(BaseModel):
    acctNum: str
    acctType: str
    name: str


class SearchCriteriaDetailModel(BaseModel):
    timePeriod: int
    txnCat: None
    viewType: str
    histSortDir: str
    acctHistSort: str
    hasBasketName: bool
    txnFromDate: str
    txnToDate: str


class GetTransactionsReqVarsModel(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    isNewOrderApi: bool
    isSupportCrypto: bool
    hideDCOrders: bool
    acctIdList: str
    acctDetailList: list[AcctDetailModel]
    searchCriteriaDetail: SearchCriteriaDetailModel


class GetTransactionsReqModel(BaseModel):
    """Fidelity get_transactions request."""

    model_config = ConfigDict(extra="forbid", strict=True)
    operationName: str
    variables: GetTransactionsReqVarsModel
    query: str


class GetTransactionsRespHistoryModel(BaseModel):
    """Single transaction in Fidelity response to get_transactions."""

    acctNum: str
    amount: str
    date: str
    description: str
    intradayInd: bool
    orderNumber: str | None


class GetTransactionsRespTransactionsModel(BaseModel):
    historys: list[GetTransactionsRespHistoryModel]


class GetTransactionsRespDataModel(BaseModel):
    getTransactions: GetTransactionsRespTransactionsModel


class GetTransactionsRespModel(BaseModel):
    """Fidelity response to get_transactions."""

    data: GetTransactionsRespDataModel


class BackendStatusModel(BaseModel):
    account: str


class SysStatusModel(BaseModel):
    backend: BackendStatusModel


class PreferenceDetailModel(BaseModel):
    name: str


class GetAccountsRespAcctDetailModel(BaseModel):
    acctNum: str
    acctType: str
    preferenceDetail: PreferenceDetailModel


class PersonModel(BaseModel):
    assets: list[GetAccountsRespAcctDetailModel]


class GetAccountsContextModel(BaseModel):
    sysStatus: SysStatusModel
    person: PersonModel


class GetAccountsRespDataModel(BaseModel):
    getContext: GetAccountsContextModel


class GetAccountsRespModel(BaseModel):
    """Fidelity response to get_accounts."""

    data: GetAccountsRespDataModel
